File: src/store.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname } from "node:path";

/**
 * Persistent map of feed URLs to sent item IDs.
 */
type StoreData = Record<string, string[]>;

/**
 * File-backed store for sent feed items.
 */
export class Store {
	/**
	 * In-memory sent item data.
	 */
	private data: StoreData = {};

	/**
	 * Creates a sent item store.
	 */
	public constructor(
		private readonly path: string,
		private readonly maxHistory: number,
	) {}

	/**
	 * Loads persisted sent item IDs if the file exists.
	 */
	public async load(): Promise<void> {
		let raw: string;
		try {
			raw = await readFile(this.path, "utf8");
		} catch (error) {
			if (isNotFound(error)) {
				return;
			}
			throw new Error(`ストア読み込み失敗: ${formatError(error)}`);
		}

		try {
			const parsed = JSON.parse(raw) as StoreData;
			this.data = normalizeStoreData(parsed);
		} catch (error) {
			throw new Error(`ストアパース失敗: ${formatError(error)}`);
		}
	}

	/**
	 * Checks whether an item was already sent for a feed URL.
	 */
	public hasSent(feedURL: string, itemID: string): boolean {
		return this.data[feedURL]?.includes(itemID) ?? false;
	}

	/**
	 * Marks an item as sent and saves the store.
	 */
	public async markSent(feedURL: string, itemID: string): Promise<void> {
		const ids = this.data[feedURL] ?? [];
		ids.push(itemID);
		this.data[feedURL] = ids.slice(-this.maxHistory);
		await this.save();
	}

	/**
	 * Saves current store data.
	 */
	private async save(): Promise<void> {
		await mkdir(dirname(this.path), { recursive: true });
		await writeFile(
			this.path,
			`${JSON.stringify(this.data, null, 2)}\n`,
			"utf8",
		);
	}
}

/**
 * Normalizes parsed store JSON into a string-array map.
 */
function normalizeStoreData(value: unknown): StoreData {
	if (!isRecord(value)) {
		return {};
	}

	const data: StoreData = {};
	for (const [key, ids] of Object.entries(value)) {
		if (Array.isArray(ids)) {
			data[key] = ids.filter((id): id is string => typeof id === "string");
		}
	}
	return data;
}

/**
 * Checks whether an fs error is ENOENT.
 */
function isNotFound(error: unknown): boolean {
	return (
		isRecord(error) &&
		"code" in error &&
		(error as { code: unknown }).code === "ENOENT"
	);
}

/**
 * Narrows unknown values to records.
 */
function isRecord(value: unknown): value is Record<string, unknown> {
	return typeof value === "object" && value !== null;
}

/**
 * Formats unknown store errors.
 */
function formatError(error: unknown): string {
	return error instanceof Error ? error.message : String(error);
}

File: internal/store/store.go
package store

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
)

type Store struct {
	mu         sync.Mutex
	data       map[string][]string
	path       string
	maxHistory int
}

func New(path string, maxHistory int) *Store {
	return &Store{
		data:       make(map[string][]string),
		path:       path,
		maxHistory: maxHistory,
	}
}

func (s *Store) Load() error {
	s.mu.Lock()
	defer s.mu.Unlock()

	raw, err := os.ReadFile(s.path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("ストア読み込み失敗: %w", err)
	}

	if err := json.Unmarshal(raw, &s.data); err != nil {
		return fmt.Errorf("ストアパース失敗: %w", err)
	}

	return nil
}

func (s *Store) HasSent(feedURL, itemID string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	ids, ok := s.data[feedURL]
	if !ok {
		return false
	}
	for _, id := range ids {
		if id == itemID {
			return true
		}
	}
	return false
}

func (s *Store) MarkSent(feedURL, itemID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.data[feedURL] = append(s.data[feedURL], itemID)

	if len(s.data[feedURL]) > s.maxHistory {
		s.data[feedURL] = s.data[feedURL][len(s.data[feedURL])-s.maxHistory:]
	}

	return s.save()
}

func (s *Store) save() error {
	dir := filepath.Dir(s.path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("ストアディレクトリ作成失敗: %w", err)
	}

	data, err := json.MarshalIndent(s.data, "", "  ")
	if err != nil {
		return fmt.Errorf("ストアのシリアライズ失敗: %w", err)
	}

	return os.WriteFile(s.path, data, 0644)
}
